server/src/main/java/org/apache/druid/discovery/BaseNodeRoleWatcher.java

```diff
@@ -59,38 +59,46 @@ public class BaseNodeRoleWatcher
   private final ConcurrentMap<String, DiscoveryDruidNode> nodes = new ConcurrentHashMap<>();
   private final Collection<DiscoveryDruidNode> unmodifiableNodes = Collections.unmodifiableCollection(nodes.values());
 
-  private final ExecutorService listenerExecutor;
+  private final ScheduledExecutorService listenerExecutor;
 
   private final List<DruidNodeDiscovery.Listener> nodeListeners = new ArrayList<>();
 
   private final Object lock = new Object();
 
+  // Always countdown under lock
   private final CountDownLatch cacheInitialized = new CountDownLatch(1);
 
+  private volatile boolean cacheInitializationTimedOut = false;
+
   public BaseNodeRoleWatcher(
-      ExecutorService listenerExecutor,
+      ScheduledExecutorService listenerExecutor,
       NodeRole nodeRole
   )
   {
-    this.listenerExecutor = listenerExecutor;
     this.nodeRole = nodeRole;
+    this.listenerExecutor = listenerExecutor;
+    this.listenerExecutor.schedule(
+        this::cacheInitializedTimedOut,
+        30L,
+        TimeUnit.SECONDS
+    );
   }
 
   public Collection<DiscoveryDruidNode> getAllNodes()
   {
+    if (cacheInitializationTimedOut) {
+      return unmodifiableNodes;
+    }
     boolean nodeViewInitialized;
     try {
-      nodeViewInitialized = cacheInitialized.await((long) 30, TimeUnit.SECONDS);
+      nodeViewInitialized = cacheInitialized.await(30L, TimeUnit.SECONDS);
```

Comment:
There's no need to wait for 30s here, since the constructor sets up a 30s timeout on its own, and ensures that the list is populated if the 30s timeout is hit. So it's better to do `await()` here and then, when the await returns, return `unmodifiableNodes`.

Reply:
Yes, removed